### PR TITLE
Add error and environment combinators for ZStream

### LINKED ISSUE #634: Generalize ZStream with input/output error types to allow for combinators like `mapError`
mapError should allow for transforming errors before running the Stream. Proposed signature:

```scala
def mapError[E2](f: E => E2): ZStream[R, E2, A]
```

### LINKED ISSUE #817: Add provide and provideSome methods to ZStream
ZIO type have a method `provide` that allows to convert `ZIO[R,E,A]` type into `ZIO[Any,E,A]`.  

It would be nice to have the same method for `ZStream`, so it can convert `ZStream[R,E,A]` type into `ZStream[Any,E,A]`.

P.S. Also we have `provideSome` method with close semantics which is also super useful.

### PR DESCRIPTION
**Pull request by iravid**
At last,
Resolves #634 
Resolves #817 

cc @runtologist @mschuwalow

### CODE REVIEW

**Comment by adamgfraser** [comment]
Yes!

**Comment by mschuwalow** [comment]
Awesome. We'll finally have this in :fireworks:

**Comment by jdegoes** [comment]
@iravid This is the look of things clicking into place. 😄 💪

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -471,6 +483,76 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
...
+      case object NotStarted extends State
+      case object Self       extends State
+      case object Other      extends State
+    }
+
+    ZStream.fromInputStreamManaged {
+      for {
+        finalizer        <- Ref.make[Exit[_, _] => ZIO[R1, Nothing, Any]](_ => UIO.unit).toManaged_
+        _                <- ZManaged.finalizerExit(e => finalizer.get.flatMap(_.apply(e)))
+        selfInputStream  <- Ref.make[InputStream[R, E, A]](InputStream.end).toManaged_
```
Comment (jdegoes) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: This pattern occurs so pervasively I think we should write a combinator to handle it (`finalizer`?). Basically creating a `Ref`, with some default action, then running some code, updating the default, and finally ensuring the finalizer is run at the end. Could be reasonably nice to use with `ZManaged`, I think.

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 2]
```diff
@@ -1132,6 +1241,61 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
       (b, leftover) = bAndLeftover
     } yield b -> (ZStream.fromChunk(leftover) ++ ZStream.fromInputStream(as))
 
+  /**
+   * Provides the stream with its required environment, which eliminates
+   * its dependency on `R`.
+   */
+  final def provide(r: R): Stream[E, A] =
```
Comment (jdegoes) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: Wonderful to see all these!

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 3]
```diff
@@ -1100,6 +1201,14 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
     }
   }
 
+  /**
+   * Switches to the provided stream in case this one fails with a typed error.
+   *
+   * See also [[ZStream#catchAll]].
+   */
+  final def orElse[R1 <: R, E2, A1 >: A](that: => ZStream[R1, E2, A1]): ZStream[R1, E2, A1] =
```
Comment (jdegoes) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: Love it!

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 4]
```diff
@@ -1476,6 +1640,30 @@ object ZStream extends ZStreamPlatformSpecific {
     def dieMessage(m: String): InputStream[Any, Nothing, Nothing]   = UIO.dieMessage(m)
     def done[E, A](e: Exit[E, A]): InputStream[Any, E, A]           = IO.done(e).mapError(Some(_))
     def fromPromise[E, A](p: Promise[E, A]): InputStream[Any, E, A] = p.await.mapError(Some(_))
+
+    def sequenceCause[E](c: Cause[Option[E]]): Option[Cause[E]] =
```
Comment (jdegoes) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: `sequenceCauseOption`?

**Comment by runtologist** [comment]
Awesome. Great job getting all those changes in. 🙏 🎉

File: streams/shared/src/main/scala/zio/stream/ZStream.scala (hunk 1 above)
Comment (iravid) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: Good idea. This looked like a combinator for `ZManaged` - `ZManaged.finalizerRef`.

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 5]
```diff
@@ -832,6 +838,24 @@ object ZManaged {
   final def finalizerExit[R](f: Exit[_, _] => ZIO[R, Nothing, Any]): ZManaged[R, Nothing, Unit] =
     ZManaged.reserve(Reservation(ZIO.unit, f))
 
+  /**
+   * Creates an effect that executes a finalizer stored in a [[Ref]]. The `Ref`
+   * is yielded as the result of the effect, allowing for control flows that require
+   * mutating finalizers.
+   */
+  final def finalizerRef[R](
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/ZManaged.scala]: This looks great!

**Pull request merged**